README.md

# rrandvec: Random Vectors that sum up to one

**Visit the [package website](https://jakobbossek.github.io/rrandvec/).**

<!-- badges: start -->

[![R-CMD-check](https://github.com/jakobbossek/rrandvec/workflows/R-CMD-check/badge.svg)](https://github.com/jakobbossek/rrandvec/actions)
[![CRAN Status
Badge](http://www.r-pkg.org/badges/version/rrandvec)](https://CRAN.R-project.org/package=rrandvec)
[![CRAN
Downloads](http://cranlogs.r-pkg.org/badges/rrandvec)](https://CRAN.R-project.org/package=rrandvec)
[![CRAN
Downloads](http://cranlogs.r-pkg.org/badges/grand-total/rrandvec?color=orange)](https://CRAN.R-project.org/package=rrandvec)
[![Codecov test
coverage](https://codecov.io/gh/jakobbossek/rrandvec/branch/main/graph/badge.svg)](https://app.codecov.io/gh/jakobbossek/rrandvec?branch=main)
<!-- badges: end -->

## What is this all about?

Offers a single method to generate random vectors whose components sum
of to one. Several methods are available. The implementation is done in
C++ (via [Rcpp](https://www.rcpp.org)) for performance reasons.

## Installation Instructions

The package can be installed from [CRAN](https://cran.r-project.org). To
install the dev-version ufrom github use the
[devtools](https://github.com/r-lib/devtools) package and type the
following command in R:

``` r
devtools::install_github("jakobbossek/rrandvec")
```

## Example

Assume we aim to generate `n=250` vectors in `d=3` dimensions where the
components of each vector follow the same distribution. This can be
achieved, e.g., by the following command:

``` r
library(rrandvec)
library(scatterplot3d)
x = rrandvec(n = 250, d = 3, method = "exponential")
scatterplot3d(x[, 1:3], angle = 120, pch = 4, color = "blue", cex.symbols = 0.7, xlab = expression(x[1]), ylab = expression(x[2]), zlab = expression(x[3]))
```

## Contact

To report bugs or ask questions please use the [issue
tracker](https://github.com/jakobbossek/rrandvec/issues). Pay attention
to explain the problem as good as possible. At its best you provide an
example, so I can reproduce your problem quickly.


NEWS.md
# rrandvec 1.0.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(rrandvec)

test_check("rrandvec")


tests/testthat/test_rpv.R
context("rrandvec")

test_that("rrandvec produces points scattered uniformly at random in the unit simplex", {
  methods = c("normalization", "trigonometric", "simplex", "iterative", "exponential")
  ds = 2:4
  n = 100L
  for (method in methods) {
    for (d in ds) {
      rnds = rrandvec(n, d, method = method, shuffle = TRUE)
      checkmate::expect_matrix(rnds, nrows = n, ncols = d, any.missing = FALSE, all.missing = FALSE)
      testthat::expect_true(all(abs(rowSums(rnds) - 1) < 0.0001))
    }
  }
})


R/rpv.R
#' @title Generate pseudo-random discrete probability distributions.
#'
#' @description Generate an \eqn{n \times d} matrix. Each row vector is a
#' probability vector \eqn{(p_1, \ldots, p_d)} with \eqn{\sum_{i=1}^{d} p_i = 1}.
#' The function offers several methods to generate the rows in a way that the
#' components are unbiased which means that they are required to have similar /
#' the same probability distributions.
#'
#' [1] Maziero, J. Generating Pseudo-Random Discrete Probability Distributions.
#' Brazilian Journal of Physics 45, 377–382 (2015).
#' https://doi.org/10.1007/s13538-015-0337-8
#'
#' [2] Grimme, C. Picking a Uniformly Random Point from an Arbitrary Simplex.
#' Technical Report. https://doi.org/10.13140/RG.2.1.3807.6968
#'
#' @param n [\code{integer(1)}]\cr
#'   Number of vectors to generate.
#' @param d [\code{integer(1)}]\cr
#'   Number of components of each vector (at least 2).
#' @param method [\code{character(1)}]\cr
#'   One of \dQuote{norm} (normalization method), \dQuote{trigonometric},
#'   \dQuote{simplex} (sample from a unit simplex), \dQuote{exponential}
#'   or \dQuote{iterative}.
#'   Default is \code{simplex}.
#' @param shuffle [\code{logical(1)}]\cr
#'   Should the values of each vector be permutatet randomly?
#'   Background: methods \dQuote{iterative} and \dQuote{trigonometric} introduce
#'   unwanted bias (see desciption). This issue can be alliviated by random
#'   shuffling.
#'   Default is \code{FALSE}.
#' @param as.df [\code{logical(1)}]\cr
#'   Should the return value be a data frame with column names X1 to Xd?
#'   Default is \code{FALSE}.
#' @return [\code{matrix(n, d)}] \eqn{(n \times d)} matrix even if \eqn{n=1}.
#' @examples
#' R = randvec(10, 2)
#' R = randvec(10, 5, method ="iterative")
#' R = randvec(10, 3, method = "trigonometric", shuffle = TRUE, as.df = TRUE)
#' \dontrun{
#' opar = par(mfrow = c(1, 3))
#' plot.ecdf(R$X1)
#' plot.ecdf(R$X2)
#' plot.ecdf(R$X3)
#' par(opar)
#' }
#' @export
randvec = function(n, d, method = "normalization", shuffle = FALSE, as.df = FALSE) {
  n = checkmate::asInt(n, lower = 1L)
  d = checkmate::asInt(d, lower = 2L)
  checkmate::assert_choice(method, choices = c("normalization", "trigonometric", "simplex", "iterative", "exponential"))
  checkmate::assert_flag(shuffle)
  checkmate::assert_flag(as.df)

  # construct name of C-method: randvec_<method>
  fun = paste0("randvec_", method)
  randvecs = do.call(fun, list(n, d))

  #FIXME: this is ugly. Could we modify the C generation code?
  if (shuffle) {
    randvecs = do.call(rbind, lapply(1:nrow(randvecs), function(i) {
      sample(randvecs[i, ])
    }))
  }

  if (as.df) {
    randvecs = as.data.frame(randvecs, stringsAsFactors = FALSE)
    colnames(randvecs) = paste0("X", 1:ncol(randvecs))
  }

  return(randvecs)
}


R/rrpv.R
#' @title Generate pseudo-random discrete probability distributions.
#'
#' @description Generate an \eqn{n \times d} matrix. Each row vector is a
#' probability vector \eqn{(p_1, \ldots, p_d)} with \eqn{\sum_{i=1}^{d} p_i = 1}.
#' The function offers several methods to generate the rows in a way that the
#' components are unbiased which means that they are required to have similar
#' probability distributions.
#'
#' @param n [\code{integer(1)}]\cr
#'   Number of vectors to generate.
#' @param d [\code{integer(1)}]\cr
#'   Number of components of each vector (at least 2).
#' @param method [\code{character(1)}]\cr
#'   One of \dQuote{norm} (normalization method), \dQuote{trigonometric},
#'   \dQuote{simplex} (sample from a unit simplex) or \dQuote{iterative}.
#'   Default is \code{simplex}.
#' @param shuffle [\code{logical(1)}]\cr
#'   Should the values of each vector be permutatet randomly?
#'   Background: methods \dQuote{iterative} and \dQuote{trigonometric} introduce
#'   unwanted bias (see desciption). This issue can be alliviated by random
#'   shuffling.
#'   Default is \code{FALSE}.
#' @param as.df [\code{logical(1)}]\cr
#'   Should the return value be a data frame with column names X1 to Xd?
#'   Default is \code{FALSE}.
#' @return [\code{matrix(n, d)}] \eqn{(n \times d)} matrix even if \eqn{n=1}.
#' @examples
#' R = rrpv(10, 2)
#' R = rrpv(10, 5, method ="iterative")
#' R = rrpv(10, 3, method = "trigonometric", shuffle = TRUE, as.df = TRUE)
#' \dontrun{
#' opar = par(mfrow = c(1, 3))
#' plot.ecdf(R$X1)
#' plot.ecdf(R$X2)
#' plot.ecdf(R$X3)
#' par(opar)
#' }
#' @export
rrpv = function(n, d, method = "normalization", shuffle = FALSE, as.df = FALSE) {
  n = checkmate::asInt(n, lower = 1L)
  d = checkmate::asInt(d, lower = 2L)
  checkmate::assert_choice(method, choices = c("normalization", "trigonometric", "simplex", "iterative"))
  checkmate::assert_flag(shuffle)
  checkmate::assert_flag(as.df)

  # construct name of C-method: rrpv_<method>_C
  fun = match.fun(paste0("rrpv_", method, "_C"))
  rpvs = fun(n, d)

  #FIXME: this is ugly. Could we modify the C generation code?
  if (shuffle) {
    rpvs = do.call(rbind, lapply(1:nrow(rpvs), function(i) {
      sample(rpvs[i, ])
    }))
  }

  if (as.df) {
    rpvs = as.data.frame(rpvs, stringsAsFactors = FALSE)
    colnames(rpvs) = paste0("X", 1:ncol(rpvs))
  }

  return(rpvs)
}


src/rpv.cpp
#include <Rcpp.h>
using namespace Rcpp;

// [[Rcpp::export]]
NumericMatrix rrandvec_normalization(int n, int d) {
  NumericMatrix R(n, d);

  for (int i = 0; i < n; ++i) {
    R(i, _) = runif(d);
  }

  NumericVector row_sums = rowSums(R);
  for (int i = 0; i < n; ++i) {
    R(i, _) = R(i, _) / row_sums[i];
  }

  return R;
}

// [[Rcpp::export]]
NumericMatrix rrandvec_iterative(int n, int d) {
  NumericMatrix R(n, d);

  for (int i = 0; i < n; ++i) {
    double sum = 0.0;
    for (int j = 0; j < d - 1; ++j) {
      R(i, j) = runif(1, 0, 1 - sum)[0];
      sum = sum + R(i, j);
    }
    R(i, d - 1) = 1 - sum;
  }

  return R;
}

// Sample a d-vector from the unit vector (see CGs technical report)
NumericVector rrandvec_sample_vector_from_unit_simplex(int d) {
  NumericVector unifs = runif(d - 1);
  std::sort(unifs.begin(), unifs.end());

  NumericVector unifs2(d + 1);
  unifs2[0] = 0;
  for (int i = 0; i < d - 1; ++i) {
    unifs2[i + 1] = unifs[i];
  }
  unifs2[d] = 1;

  NumericVector out(d);
  for (int i = 1; i <= d; ++i) {
    out[i - 1] = unifs2[i] - unifs2[i - 1];
  }
  return out;
}

// [[Rcpp::export]]
NumericMatrix rrandvec_simplex(int n, int d) {
  NumericMatrix R(n, d);

  for (int i = 0; i < n; ++i) {
    R(i, _) = rrandvec_sample_vector_from_unit_simplex(d);
  }

  return R;
}

// [[Rcpp::export]]
NumericMatrix rrandvec_trigonometric(int n, int d) {
  NumericMatrix R(n, d);

  for (int i = 0; i < n; ++i) {
    NumericVector ts = runif(d - 1);
    NumericVector thetas(d);
    thetas[0] = 3.1415 / 2; // pi/2
    for (int j = d - 1; j > 0; --j) {
      thetas[j] = acos(sqrt(ts[j - 1]));
    }

    for (int j = d - 1; j >= 0; --j) {
      double r = 0.0;
      r = sin(thetas[j]) * sin(thetas[j]);
      for (int k = j + 1; k < d; ++k) {
        r = r * cos(thetas[k]) * cos(thetas[k]);
      }
      R(i, j) = r;
    }
  }

  return R;
}

// [[Rcpp::export]]
NumericMatrix rrandvec_exponential(int n, int d) {
  NumericMatrix R(n, d);

  double u, s;
  for (int i = 0; i < n; ++i) {
    s = 0;
    for (int j = 0; j < d; ++j) {
      u = as<double>(runif(1)); // ((double) rand() / (RAND_MAX));
      R(i, j) = (-1) * log(1 - u);
      s += R(i, j);
    }
    for (int j = 0; j < d; ++j) {
      R(i, j) = R(i, j) / s;
    }
  }

  return R;
}


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

#ifdef RCPP_USE_GLOBAL_ROSTREAM
Rcpp::Rostream<true>&  Rcpp::Rcout = Rcpp::Rcpp_cout_get();
Rcpp::Rostream<false>& Rcpp::Rcerr = Rcpp::Rcpp_cerr_get();
#endif

// rrandvec_normalization
NumericMatrix rrandvec_normalization(int n, int d);
RcppExport SEXP _rrandvec_rrandvec_normalization(SEXP nSEXP, SEXP dSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< int >::type n(nSEXP);
    Rcpp::traits::input_parameter< int >::type d(dSEXP);
    rcpp_result_gen = Rcpp::wrap(rrandvec_normalization(n, d));
    return rcpp_result_gen;
END_RCPP
}
// rrandvec_iterative
NumericMatrix rrandvec_iterative(int n, int d);
RcppExport SEXP _rrandvec_rrandvec_iterative(SEXP nSEXP, SEXP dSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< int >::type n(nSEXP);
    Rcpp::traits::input_parameter< int >::type d(dSEXP);
    rcpp_result_gen = Rcpp::wrap(rrandvec_iterative(n, d));
    return rcpp_result_gen;
END_RCPP
}
// rrandvec_simplex
NumericMatrix rrandvec_simplex(int n, int d);
RcppExport SEXP _rrandvec_rrandvec_simplex(SEXP nSEXP, SEXP dSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< int >::type n(nSEXP);
    Rcpp::traits::input_parameter< int >::type d(dSEXP);
    rcpp_result_gen = Rcpp::wrap(rrandvec_simplex(n, d));
    return rcpp_result_gen;
END_RCPP
}
// rrandvec_trigonometric
NumericMatrix rrandvec_trigonometric(int n, int d);
RcppExport SEXP _rrandvec_rrandvec_trigonometric(SEXP nSEXP, SEXP dSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< int >::type n(nSEXP);
    Rcpp::traits::input_parameter< int >::type d(dSEXP);
    rcpp_result_gen = Rcpp::wrap(rrandvec_trigonometric(n, d));
    return rcpp_result_gen;
END_RCPP
}
// rrandvec_exponential
NumericMatrix rrandvec_exponential(int n, int d);
RcppExport SEXP _rrandvec_rrandvec_exponential(SEXP nSEXP, SEXP dSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< int >::type n(nSEXP);
    Rcpp::traits::input_parameter< int >::type d(dSEXP);
    rcpp_result_gen = Rcpp::wrap(rrandvec_exponential(n, d));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_rrandvec_rrandvec_normalization", (DL_FUNC) &_rrandvec_rrandvec_normalization, 2},
    {"_rrandvec_rrandvec_iterative", (DL_FUNC) &_rrandvec_rrandvec_iterative, 2},
    {"_rrandvec_rrandvec_simplex", (DL_FUNC) &_rrandvec_rrandvec_simplex, 2},
    {"_rrandvec_rrandvec_trigonometric", (DL_FUNC) &_rrandvec_rrandvec_trigonometric, 2},
    {"_rrandvec_rrandvec_exponential", (DL_FUNC) &_rrandvec_rrandvec_exponential, 2},
    {NULL, NULL, 0}
};

RcppExport void R_init_rrandvec(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/rrandvec.R
#' @title Generate random vectors that sum up to one.
#'
#' @description Generate an \eqn{n \times d} matrix. Each row vector is a
#' probability vector \eqn{(p_1, \ldots, p_d)} with \eqn{\sum_{i=1}^{d} p_i = 1}.
#' The function offers several methods to generate the rows in a way that the
#' components are unbiased which means that they are required to have similar /
#' the same probability distributions.
#'
#' [1] Maziero, J. Generating Pseudo-Random Discrete Probability Distributions.
#' Brazilian Journal of Physics 45, 377–382 (2015).
#' https://doi.org/10.1007/s13538-015-0337-8
#'
#' [2] Grimme, C. Picking a Uniformly Random Point from an Arbitrary Simplex.
#' Technical Report. https://doi.org/10.13140/RG.2.1.3807.6968
#'
#' @param n [\code{integer(1)}]\cr
#'   Number of vectors to generate.
#' @param d [\code{integer(1)}]\cr
#'   Number of components of each vector (at least 2).
#' @param method [\code{character(1)}]\cr
#'   One of \dQuote{norm} (normalization method), \dQuote{trigonometric},
#'   \dQuote{simplex} (sample from a unit simplex), \dQuote{exponential}
#'   or \dQuote{iterative}.
#'   Default is \code{simplex}.
#' @param shuffle [\code{logical(1)}]\cr
#'   Should the values of each vector be permutatet randomly?
#'   Background: methods \dQuote{iterative} and \dQuote{trigonometric} introduce
#'   unwanted bias (see desciption). This issue can be alliviated by random
#'   shuffling.
#'   Default is \code{FALSE}.
#' @param as.df [\code{logical(1)}]\cr
#'   Should the return value be a data frame with column names X1 to Xd?
#'   Default is \code{FALSE}.
#' @return [\code{matrix(n, d)}] \eqn{(n \times d)} matrix even if \eqn{n=1}.
#' @examples
#' R = rrandvec(1000, 2)
#' R = rrandvec(1000, 5, method ="iterative")
#' R = rrandvec(1000, 3, method = "trigonometric", shuffle = TRUE, as.df = TRUE)
#'
#' if (require("scatterplot3d")) {
#'   scatterplot3d::scatterplot3d(R, angle = 120, cex.symbols = 0.5, pch = 3, color = "blue")
#' }
#' @export
rrandvec = function(n, d, method = "normalization", shuffle = FALSE, as.df = FALSE) {
  n = checkmate::asInt(n, lower = 1L)
  d = checkmate::asInt(d, lower = 2L)
  checkmate::assert_choice(method, choices = c("normalization", "trigonometric", "simplex", "iterative", "exponential"))
  checkmate::assert_flag(shuffle)
  checkmate::assert_flag(as.df)

  # construct name of C-method: rrandvec_<method>
  fun = paste0("rrandvec_", method)
  rrandvecs = do.call(fun, list(n, d))

  #FIXME: this is ugly. Could we modify the C generation code?
  if (shuffle) {
    rrandvecs = do.call(rbind, lapply(1:nrow(rrandvecs), function(i) {
      sample(rrandvecs[i, ])
    }))
  }

  if (as.df) {
    rrandvecs = as.data.frame(rrandvecs, stringsAsFactors = FALSE)
    colnames(rrandvecs) = paste0("X", 1:ncol(rrandvecs))
  }

  return(rrandvecs)
}
